parser-typechecker/src/Unison/Syntax/TermParser.hs

```diff
@@ -1106,66 +1113,69 @@ substImports ns imports =
   String ->
   P v (L.Token ()) ->
-  P v b ->
+  P v (L.Token ()) ->
   TermP v
 block'' isTop implicitUnitAtEnd s openBlock closeBlock = do
   open <- openBlock
   (names, imports) <- imports
   _ <- optional semi
   statements <- local (\e -> e {names = names}) $ sepBy semi statement
-  _ <- closeBlock
-  substImports names imports <$> go open statements
+  _close <- closeBlock
+  theBlock <- substImports names imports <$> go open statements
+  pure $ ABT.annotate (ann open <> ann theBlock) theBlock
   where
     statement = asum [Binding <$> binding, DestructuringBind <$> destructuringBind, Action <$> blockTerm]
     go :: L.Token () -> [BlockElement v] -> P v (Term v Ann)
     go open bs =
-      let finish tm = case Components.minimize' tm of
+      let finish :: Term.Term v Ann -> TermP v
+          finish tm = case Components.minimize' tm of
             Left dups -> customFailure $ DuplicateTermNames (toList dups)
             Right tm -> pure tm
-          toTm bs = do
-            (bs, body) <- body bs
-            finish =<< foldrM step body bs
+          toTm :: [BlockElement v] -> TermP v
+          toTm [] = customFailure $ EmptyBlock (const s <$> open)
+          toTm (be : bes) = do
+            let (bs, blockResult) = determineBlockResult (be :| bes)
+            finish =<< foldrM step blockResult bs
             where
-              step elem body = case elem of
+              step :: BlockElement v -> Term v Ann -> TermP v
+              step elem result = case elem of
                 Binding ((a, v), tm) ->
                   pure $
                     Term.consLetRec
                       isTop
-                      (ann a <> ann body)
+                      (ann a <> ann result)
                       (a, v, tm)
-                      body
+                      result
                 Action tm ->
                   pure $
                     Term.consLetRec
                       isTop
-                      (ann tm <> ann body)
+                      (ann tm <> ann result)
                       (ann tm, positionalVar (ann tm) (Var.named "_"), tm)
-                      body
+                      result
                 DestructuringBind (_, f) ->
-                  f <$> finish body
-          body bs = case reverse bs of
-            Binding ((a, _v), _) : _ ->
-              pure $
-                if implicitUnitAtEnd
-                  then (bs, DD.unitTerm a)
-                  else (bs, Term.var a (positionalVar a Var.missingResult))
-            Action e : bs -> pure (reverse bs, e)
-            DestructuringBind (a, _) : _ ->
-              pure $
-                if implicitUnitAtEnd
-                  then (bs, DD.unitTerm a)
-                  else (bs, Term.var a (positionalVar a Var.missingResult))
-            [] -> customFailure $ EmptyBlock (const s <$> open)
+                  f <$> finish result
+          determineBlockResult :: NonEmpty (BlockElement v) -> ([BlockElement v], Term v Ann)
```

Comment:
These changes are just cosmetic, I moved the monad up outside of the combinator and added some types and such.